python/pyarrow/filesystem.py

```diff
@@ -251,6 +251,10 @@ def isfile(self, path):
     def delete(self, path, recursive=False):
         return self.fs.rm(path, recursive=recursive)
 
+    @implements(FileSystem.exists)
+    def exists(self, path):
+        return os.path.exists(path)
```

Comment:
Does this need to be `self.fs.exists(path)`?

Reply:
@wesm yes 🤦‍♂️

Reply:
I wrote my own wrapper to get around this issue and I realised that pyarrow would try and create a directory if 'exists' returns False. This is an issue on s3 since there's no concept of directories. So what ends up happening is that there's a duplicate file with the same name as the directory.
You almost need to remove the `exists()` check for s3 filesystem.

Reply:
Yeah, I did not implement exists because of this issue. This needs to be tested against S3 before being merged

Reply:
Oh thats interesting @johnjiang 

I will test this with S3 as @wesm suggested and see if I can come up with a solution that is satisfactory.

Reply:
@wesm @johnjiang I added a new attribute to `FileSystem` which simply indicates whether the FileSystem is an object-store or a file-store. Then added an additional param to the mkdir conditional in `write_to_dataset` to check if the `FileSystem` is a File store before attempting to call `mkdir`. I imagine that tiny method could come in handy to distinguish other features between object/file stores later. Let me know if that addresses your concerns, otherwise happy to try something else out.